### ISSUE TITLE
Training Cost NAN

### ISSUE THREAD

**Issue by jiangqy** [uoguelph-mlrg/theano_alexnet#27]
Hi, I would like to train AlexNet on ImageNet. While after 20 iterations, training cost becomes nan. 
Here are the details:

![d8yaw cmo5 ynyr t_ krs](https://cloud.githubusercontent.com/assets/14566109/14585080/24d43bfa-0499-11e6-8143-5ea9ed09a167.png)

Should I set a smaller learning rate? Could you give me some suggestions?

Thank you~

**Comment by hma02** [comment]
@jiangqy 
What is your batch size and current learning rate?

**Comment by heipangpang** [comment]
I have met the same problem, and my batch size is 256, learning rate is 0.01. Do you have any ideas?

**Comment by jiangqy** [comment]
@hma02 My batch size is 256 and learning rate is 0.01, too.

**Comment by hma02** [comment]
@jiangqy @heipangpang 
Looks like you are running the single GPU train.py, then the problem is not related to weight exchanging. 

The cost should be around 6.9 initially.

The unbounded cost value maybe caused by gradient explosion. I got into similar situations when initializing a deep network with arrays of large variance and mean. Too large learning rates and batch sizes may result in strong gradient zigzag as well.

Also do check the input images when [loading them](https://github.com/uoguelph-mlrg/theano_alexnet/blob/master/proc_load.py#L119)
 to see if they are preprocessed correctly and correspond to loaded labels
You can show them using similar tricks as [here](https://github.com/hma02/show_batch/blob/master/show_batch.py#L36). Try using a stack of image_means as [input data](https://github.com/uoguelph-mlrg/theano_alexnet/blob/master/proc_load.py#L119)

**Comment by heipangpang** [comment]
@hma02 
I will try it.  Thank you very much.

**Comment by heipangpang** [comment]
@hma02 
When I check the output of every layer, I found that for the layer_input, I got a zero matrix which may be the problem why I such a large training loss.

**Comment by hma02** [comment]
@heipangpang 
Yes, this probably is the reason you got large cost. Make sure you set `use_data_layer` to False in config.yaml. Then the layer_input should be equal to x as shown [here](https://github.com/uoguelph-mlrg/theano_alexnet/blob/master/alex_net.py#L42), which is the input batch. If x is a zero matrix, there's something wrong with the preprocessed image batches.

**Comment by heipangpang** [comment]
@hma02 
But when I load the batches hand by hand in python, it seems that I can get the correct results.
Thank you very much.

**Comment by heipangpang** [comment]
@hma02 
I am getting the correct results now, thank you very much.

**Comment by liaocs2008** [comment]
The code didn't load data in parallel. Check the function train_model_wrap().

**Comment by liaocs2008** [comment]
I had the same problem here. If "para_load" is set False, I could train it normally. But I think one of great contributions of this work is the parallel loading right?

**Comment by Magotraa** [comment]
@heipangpang
Can you please share, what change exactly made it possible for you to get correct results. 

As you wrote
"I am getting the correct results now, thank you very much."
